substratevm/src/com.oracle.svm.agent.core/src/com/oracle/svm/agent/core/JNIHandleSetBase.java

```diff
@@ -0,0 +1,176 @@
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package com.oracle.svm.agent.core;
+
+import static com.oracle.svm.core.util.VMError.guarantee;
+import static com.oracle.svm.jni.JNIObjectHandles.nullHandle;
+
+import java.util.Arrays;
+import java.util.concurrent.locks.ReentrantLock;
+
+import org.graalvm.nativeimage.c.type.CTypeConversion;
+
+import com.oracle.svm.jni.JNIObjectHandles;
+import com.oracle.svm.jni.nativeapi.JNIEnvironment;
+import com.oracle.svm.jni.nativeapi.JNIMethodId;
+import com.oracle.svm.jni.nativeapi.JNIObjectHandle;
+
+/**
+ * Helps with creation and manages JNI handles for JVMTI agents.
+ *
+ * A JVMTI agent must provide a subclass of this class. It should contain handles to classes and
+ * methods that are needed across different JVMTI calls (for example, a JNIObjectHandle to
+ * java/lang/Class).
+ *
+ * For JNI handles that are created in a JVMTI callback and that should survive a return from native
+ * code, a helper method {@link #newTrackedGlobalRef} is provided.
+ *
+ * See also: {@link com.oracle.svm.agent.core.JvmtiAgentBase}.
+ */
+public abstract class JNIHandleSetBase {
+    private static final int INITIAL_GLOBAL_HANDLE_CAPACITY = 16;
+    private final ReentrantLock globalRefsLock = new ReentrantLock();
+    private JNIObjectHandle[] globalRefs = new JNIObjectHandle[INITIAL_GLOBAL_HANDLE_CAPACITY];
+    private int globalRefCount = 0;
+
+    JNIMethodId javaLangClassGetName;
+
+    public JNIHandleSetBase(JNIEnvironment env) {
+        JNIObjectHandle javaLangClass = findClass(env, "java/lang/Class");
+        try (CTypeConversion.CCharPointerHolder name = Support.toCString("getName"); CTypeConversion.CCharPointerHolder signature = Support.toCString("()Ljava/lang/String;")) {
+            javaLangClassGetName = Support.jniFunctions().getGetMethodID().invoke(env, javaLangClass, name.get(), signature.get());
+            guarantee(javaLangClassGetName.isNonNull());
+        }
+    }
+
+    /**
+     * Returns a local handle to a Java class object.
+     *
+     * If the class is not found, a RuntimeException is thrown. If the handle must survive a return
+     * from native code, {@link #newTrackedGlobalRef} should be used on the returned handle.
```

Comment:
`guarantee` at runtime is substituted and aborts the entire VM, see the substitution in `Target_com_oracle_svm_core_util_VMError.shouldNotReachHere`. The class _must_ exist.

Reply:
I agree, the substitution changes the exception and that exception cannot be caught. I will update the docs.